include/boost/numeric/odeint/stepper/controlled_adams_bashforth_moulton.hpp

```diff
@@ -39,39 +39,54 @@ class default_order_adjuster
     : m_algebra( algebra )
     {};
 
-    void adjust_order(size_t &order, const boost::array<wrapped_state_type, 3> & xerr)
+    void adjust_order(size_t &order, size_t init, boost::array<wrapped_state_type, 4> &xerr)
     {
         using std::abs;
 
-        value_type errm = abs(m_algebra.norm_inf(xerr[0].m_v));
-        value_type errc = abs(m_algebra.norm_inf(xerr[1].m_v));
-        value_type errp = abs(m_algebra.norm_inf(xerr[2].m_v));
+        value_type errc = abs(m_algebra.norm_inf(xerr[2].m_v));
 
-        if(order == 1)
-        {   
-            if(errp < errc)
-            {
-                order ++;
-            }
+        value_type errm1 = 3*errc;
+        value_type errm2 = 3*errc;
+
+        if(order > 2)
+        {
+            errm2 = abs(m_algebra.norm_inf(xerr[0].m_v));
         }
-        else if(order == MaxOrder)
+        if(order >= 2)
         {
-            if(errm < errc)
-            {
-                order --;
-            }
+            errm1 = abs(m_algebra.norm_inf(xerr[1].m_v));
         }
-        else
+
+        size_t o_new = order;
+
+        if(order == 2 && errm1 <= 0.5*errc)
+        {
+            o_new = order - 1;
+        }
+        else if(order > 2 && errm2 < errc && errm1 < errc)
+        {
+            o_new = order - 1;
+        }
+
+        if(init < order)
+        {
+            ++order;
```

Comment:
I don't really understand what's going on but this looks a bit suspicious. If order > init: order++ seems odd - please verify this is correct.

Reply:
This might be a ambiguous because of the variable name 'init'. 'init' is the number of currently initialized steps in the memory. What happens here is that the order is automatically increased if the error for the order above the current one can not be calculated yet. 

This way of choosing the order seems to be standard among the codes I looked at.